ablateLibrary/particles/particles.cpp

```diff
@@ -544,25 +544,108 @@ static PetscErrorCode DMSequenceViewTimeHDF5(DM dm, PetscViewer viewer) {
     PetscFunctionReturn(0);
 }
 
-void ablate::particles::Particles::View(PetscViewer viewer, PetscInt steps, PetscReal time, Vec u) const {
+void ablate::particles::Particles::Save(PetscViewer viewer, PetscInt steps, PetscReal time) const {
     DMSetOutputSequenceNumber(GetDM(), steps, time) >> checkError;
-
-    PetscFunctionBegin;
     Vec particleVector;
 
     for (auto const &field : particleFieldDescriptors) {
-        if (field.type == PETSC_DOUBLE) {
+        if (field.type == PETSC_REAL) {
             DMSwarmCreateGlobalVectorFromField(GetDM(), field.fieldName.c_str(), &particleVector) >> checkError;
             PetscObjectSetName((PetscObject)particleVector, field.fieldName.c_str()) >> checkError;
             VecView(particleVector, viewer) >> checkError;
             DMSwarmDestroyGlobalVectorFromField(GetDM(), field.fieldName.c_str(), &particleVector) >> checkError;
         }
     }
 
+    // Get the particle info
+    int rank;
+    MPI_Comm_rank(PetscObjectComm((PetscObject)GetDM()), &rank) >> checkMpiError;
+
+    // get the local number of particles
+    PetscInt localSize;
+    DMSwarmGetLocalSize(GetDM(), &localSize) >> checkMpiError;
+    ;
+
+    // record the number of particles per rank
+    Vec particlesPerRank;
+    VecCreateMPI(PETSC_COMM_WORLD, 1, PETSC_DECIDE, &particlesPerRank) >> checkError;
+    PetscObjectSetName((PetscObject)particlesPerRank, "particlesPerRank") >> checkError;
+    VecSetValue(particlesPerRank, rank, localSize, INSERT_VALUES) >> checkError;
+    VecAssemblyBegin(particlesPerRank) >> checkError;
+    VecAssemblyEnd(particlesPerRank) >> checkError;
+    VecView(particlesPerRank, viewer);
+    VecDestroy(&particlesPerRank) >> checkError;
+
     // if this is an hdf5Viewer
     PetscBool ishdf5;
     PetscObjectTypeCompare((PetscObject)viewer, PETSCVIEWERHDF5, &ishdf5) >> checkError;
     if (ishdf5) {
         DMSequenceViewTimeHDF5(GetDM(), viewer) >> checkError;
     }
 }
+
```

Comment:
@knepley Do you know if there is a better way to save/restore to dmSwarm?

Reply:
Currently, no. You are right that we should move this code into DMView()/DMLoad()